src/stacked_borrows/stack.rs

```diff
@@ -0,0 +1,307 @@
+}
+
+#[cfg(feature = "stack-cache")]
+impl StackCache {
+    fn add(&mut self, idx: usize, tag: SbTag) {
+        self.tags.copy_within(0..CACHE_LEN - 1, 1);
+        self.tags[0] = tag;
+        self.idx.copy_within(0..CACHE_LEN - 1, 1);
+        self.idx[0] = idx;
+    }
+}
+
+impl PartialEq for Stack {
+    fn eq(&self, other: &Self) -> bool {
+        // All the semantics of Stack are in self.borrows, everything else is caching
+        self.borrows == other.borrows
+    }
+}
+
+impl Eq for Stack {}
+
+impl<'tcx> Stack {
+    /// Panics if any of the caching mechanisms have broken,
+    /// - The StackCache indices don't refer to the parallel tags,
+    /// - There are no Unique tags outside of first_unique..last_unique
+    #[cfg(feature = "expensive-debug-assertions")]
+    fn verify_cache_consistency(&self) {
+        for (tag, stack_idx) in self.cache.tags.iter().zip(self.cache.idx.iter()) {
+            assert_eq!(self.borrows[*stack_idx].tag, *tag);
+        }
+
+        for (idx, item) in self.borrows.iter().enumerate() {
+            if item.perm == Permission::Unique {
+                assert!(
+                    self.unique_range.contains(&idx),
+                    "{:?} {:?}",
+                    self.unique_range,
+                    self.borrows
+                );
+            }
+        }
+    }
+
+    /// Find the item granting the given kind of access to the given tag, and return where
+    /// it is on the stack.
+    pub fn find_granting(&mut self, access: AccessKind, tag: SbTag) -> Option<usize> {
+        #[cfg(feature = "expensive-debug-assertions")]
+        self.verify_cache_consistency();
+
+        match tag {
+            SbTag::Untagged =>
+                self.borrows.iter().enumerate().rev().find_map(|(idx, item)| {
+                    if item.tag == SbTag::Untagged && item.perm.grants(access) {
+                        Some(idx)
+                    } else {
+                        None
+                    }
+                }),
+            SbTag::Tagged(_) => {
+                #[cfg(feature = "stack-cache")]
+                if let Some(stack_idx) = self.find_granting_cache(access, tag) {
+                    return Some(stack_idx);
+                }
+                // If we didn't find the tag in the cache, fall back to a linear search of the
+                // whole stack, and add the tag to the stack.
+                for (stack_idx, item) in self.borrows.iter().enumerate().rev() {
+                    if tag == item.tag && item.perm.grants(access) {
+                        #[cfg(feature = "stack-cache")]
+                        self.cache.add(stack_idx, tag);
+                        return Some(stack_idx);
+                    }
+                }
+                None
+            }
+        }
+    }
+
+    #[cfg(feature = "stack-cache")]
+    fn find_granting_cache(&mut self, access: AccessKind, tag: SbTag) -> Option<usize> {
+        if self.borrows.len() > CACHE_LEN {
```

Comment:
You can avoid some rightwards drift here by negating the condition and doing an early `return None`. Also please add a comment explaining the purpose of this.